### PR TITLE
Issue 381: Relax actor_type to optional.

### LINKED ISSUE #381: Actor: make actor_type optional.
`actor_type` [is required](https://github.com/threatgrid/ctim/blob/master/doc/structures/actor.md), but we want to ingest Mitre ATT&CK Groups, which don't include data to populate this field. [E.g.](https://github.com/mitre/cti/blob/master/enterprise-attack/intrusion-set/intrusion-set--00f67a77-86a4-4adf-be26-1a54fc713340.json)

Is it safe to make this field optional? We know that public CTIA does not currently have any Actors loaded.

### PR DESCRIPTION
**Pull request by samwagg**
closes #381

### CODE REVIEW

File: src/ctim/schemas/actor.cljc [hunk 1]
```diff
@@ -23,9 +23,9 @@
   c/described-entity-entries
   (f/required-entries
    (f/entry :type ActorTypeIdentifier)
-   (f/entry :valid_time c/ValidTime)
-   (f/entry :actor_type v/ThreatActorType))
+   (f/entry :valid_time c/ValidTime))
   (f/optional-entries
+   (f/entry :actor_type v/ThreatActorType)
```
Comment (msprunck) [inline src/ctim/schemas/actor.cljc:28]: What about changing it to a collection as I suggested in this https://github.com/threatgrid/ctim/issues/381#issuecomment-1314889560?

Comment (ereteog) [reply src/ctim/schemas/actor.cljc:28]: if there are any in prod today, we will have to migrate it. But yes that would be nice (and that's an easy one to migrate)

**Comment by samwagg** [comment]
I think this is ready now.

**Review by msprunck — APPROVED** [review]
LGTM! Could you check if we have Actor entities in PROD? If yes, we probably need to perform a migration. Could you also change the schema version [here](https://github.com/threatgrid/ctim/blob/master/src/ctim/schemas/common.cljc#L23)?

**Comment by samwagg** [comment]
> LGTM! Could you check if we have Actor entities in PROD? If yes, we probably need to perform a migration. Could you also change the schema version [here](https://github.com/threatgrid/ctim/blob/master/src/ctim/schemas/common.cljc#L23)?

1. There are no actor entities in public CTIA in prod. I don't know much about private CTIAs and whether we need to consider those.

2. Can you advise on the version level to bump? I would be inclined to up the major version, given this would require a migration.

**Comment by samwagg** [comment]
Private CTIA does have Actors. 268 of them. So I guess a migration is needed.

```
curl -X GET "https://es-private-storage.us-east-1.prod.iroh.site/v1.1.0_private_ctia_actor-2019.11.19-000001/_search?pretty" -H 'Content-Type: application/json' -d'
{
  "query": {
    "match_all": { }
  } 
}
'
```

**Comment by samwagg** [comment]
Query of Actor index on private CTIA returned 268 hits. So I guess a migration is needed.

**Comment by ereteog** [comment]
that's very old data and the non garbarge source is `threatq`, the most recent data being from 2020.
I made this aggregation directly on the cluster:
``` json
{
    "aggs": {
        "groups": {
            "terms": {
                "field": "groups",
                "size": 100
            },
            "aggs": {
                "sources": {
                    "terms": {
                        "field": "source"
                    }
                },
                "year": {
                    "date_histogram": {
                        "field": "created",
                        "calendar_interval": "year"
                    }
                }
            }
        }
    }
}
```

<details>

<summary> results </summary>
```
{
    "took": 14,
    "timed_out": false,
    "_shards": {
        "total": 5,
        "successful": 5,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 258,
            "relation": "eq"
        },
        "max_score": null,
        "hits": []
    },
    "aggregations": {
        "groups": {
            "doc_count_error_upper_bound": 0,
            "sum_other_doc_count": 0,
            "buckets": [
                {
                    "key": "bcde1e51-5020-45b8-ba11-f790b0abb8b8",
                    "doc_count": 148,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "threatq",
                                "doc_count": 148
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2019-01-01T00:00:00.000Z",
                                "key": 1546300800000,
                                "doc_count": 95
                            },
                            {
                                "key_as_string": "2020-01-01T00:00:00.000Z",
                                "key": 1577836800000,
                                "doc_count": 53
                            }
                        ]
                    }
                },
                {
                    "key": "ea3f7333-aede-44b2-b3d5-0a9e0e1b52a2",
                    "doc_count": 65,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "a source",
                                "doc_count": 56
                            },
                            {
                                "key": "string",
                                "doc_count": 4
                            },
                            {
                                "key": "test source",
                                "doc_count": 3
                            },
                            {
                                "key": "new source point",
                                "doc_count": 2
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2019-01-01T00:00:00.000Z",
                                "key": 1546300800000,
                                "doc_count": 46
                            },
                            {
                                "key_as_string": "2020-01-01T00:00:00.000Z",
                                "key": 1577836800000,
                                "doc_count": 2
                            },
                            {
                                "key_as_string": "2021-01-01T00:00:00.000Z",
                                "key": 1609459200000,
                                "doc_count": 17
                            }
                        ]
                    }
                },
                {
                    "key": "threatgrid:12882",
                    "doc_count": 28,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "threatq",
                                "doc_count": 28
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2020-01-01T00:00:00.000Z",
                                "key": 1577836800000,
                                "doc_count": 28
                            }
                        ]
                    }
                },
                {
                    "key": "threatgrid:200818",
                    "doc_count": 9,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "threatq",
                                "doc_count": 9
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2020-01-01T00:00:00.000Z",
                                "key": 1577836800000,
                                "doc_count": 9
                            }
                        ]
                    }
                },
                {
                    "key": "threatgrid:1",
                    "doc_count": 3,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "string",
                                "doc_count": 3
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2020-01-01T00:00:00.000Z",
                                "key": 1577836800000,
                                "doc_count": 3
                            }
                        ]
                    }
                },
                {
                    "key": "18fc4fd9-fb31-48bd-a828-a8a0783a1d2e",
                    "doc_count": 1,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "threatq",
                                "doc_count": 1
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2020-01-01T00:00:00.000Z",
                                "key": 1577836800000,
                                "doc_count": 1
                            }
                        ]
                    }
                },
                {
                    "key": "97081d18-f1b9-4c75-9e23-cda8f7bb0368",
                    "doc_count": 1,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "string",
                                "doc_count": 1
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2021-01-01T00:00:00.000Z",
                                "key": 1609459200000,
                                "doc_count": 1
                            }
                        ]
                    }
                },
                {
                    "key": "c5ac236d-5352-42ff-99b1-f6dd792fb1f1",
                    "doc_count": 1,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "test source",
                                "doc_count": 1
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2021-01-01T00:00:00.000Z",
                                "key": 1609459200000,
                                "doc_count": 1
                            }
                        ]
                    }
                },
                {
                    "key": "fake-org-id",
                    "doc_count": 1,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "string",
                                "doc_count": 1
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2019-01-01T00:00:00.000Z",
                                "key": 1546300800000,
                                "doc_count": 1
                            }
                        ]
                    }
                },
                {
                    "key": "threatgrid:1242",
                    "doc_count": 1,
                    "sources": {
                        "doc_count_error_upper_bound": 0,
                        "sum_other_doc_count": 0,
                        "buckets": [
                            {
                                "key": "string",
                                "doc_count": 1
                            }
                        ]
                    },
                    "year": {
                        "buckets": [
                            {
                                "key_as_string": "2018-01-01T00:00:00.000Z",
                                "key": 1514764800000,
                                "doc_count": 1
                            }
                        ]
                    }
                }
            ]
        }
    }
}
```
</details>

**Comment by ereteog** [comment]
@samwagg @michaels-den for the next steps, you need to make a CTIA PR with:
- an update of the mapping here: https://github.com/threatgrid/ctia/blob/f30a0046ff2e07b7e82f7664604ae28edd9d2403/src/ctia/entity/actor.clj#L48
- a function that performs the transformation of a document like: https://github.com/threatgrid/ctia/blob/f30a0046ff2e07b7e82f7664604ae28edd9d2403/src/ctia/task/migration/migrations.clj#L171
- a bump of CTIM version

**Comment by samwagg** [comment]
I wanted to make sure I more or less understood how migrations work before proceeding. I think I do now. Please take a glance at my draft PR at https://github.com/threatgrid/ctia/pull/1323 and make sure it seems okay and then I'll start moving forward on merging this and publishing to clojars.

**Comment by ereteog** [comment]
Yes that's what you have to implement.
Once done, you need to prepare the tenzin configuration to trigger the migration: https://github.com/threatgrid/ctia/blob/master/resources/ctia/public/doc/migration.md

**Comment by samwagg** [comment]
> Rebased to resolve merge conflicts. This is ready for merge @ereteog or @msprunck

whoops, forgot to do the version change. One sec...

**Comment by samwagg** [comment]
Bumped version to 1.2.0 and squashed commits. _Now_ I think it's ready for merge.

**Pull request merged**